# <center> 🐼 [Введение в Pandas](https://stepik.org/lesson/681964/)</center>

Этот ноутбук поможет разобраться вам в основных моментах предварительной обработки данных в Pandas, познакомиться с полезными функциями и трюками. 

### Оглавление ноутбука
<img src="https://sun2.beeline-kz.userapi.com/impg/sH8DAKXIRcSQkiZ119iicN4-8JdSEjH88T1QfQ/53wjU4ABcrI.jpg?size=728x455&quality=96&sign=caed6c5a44eb5b193ed9d8e975558c56&type=album" align="right" width="528" height="528" />
<br>

<p><font size="3" face="Arial" font-size="large">
<ul type="square"><li><a href="#1">📑 Структуры данных в Pandas</a></li><li><a href="#2">🔍 Разведочный анализ данных</a><ul><li><a href="#3">Загрузка и первичное знакомство</a></li><li><a href="#4">Предварительный осмотр</a></li><li><a href="#5">Работа с числовыми данными</a></li></ul></li><li><a href="#6">📤 Извлечение необходимой информации</a><ul><li><a href="#7">Сортировка</a></li><li><a href="#8">Методы фильтрации</a></li><li><a href="#9">Работа с пропусками</a></li><li><a href="#10">Индексация данных</a></li><li><a href="#11">Функции</a></li></ul></li>
</ul></font></p>

# <center>📑 Cтруктуры данных в Pandas. </center>

<p id="1"></p>

`Series` — это структура, используемая для работы с последовательностью одномерных данных, а `Dataframe` — более сложная и подходит для нескольких измерений. `Series` – одномерный массив с метками, способный хранить любой тип данных.

<img src="https://sun9-43.userapi.com/impg/0k7fzcgju-6hY10XmQoaJkfxOxr0-SvgA3NHQg/N0Hb-T_fXio.jpg?size=1280x720&quality=96&sign=6ac22e0ff168023b0dff780c1bfa5e36&type=album"></img>

In [1]:
import pandas as pd # pandas для работы с табличными данными
import numpy as np  # numpy для математических действий

### Создать`Series`  можно так:

`s = pd.Series(data, index=index, name='something')`

Метки именуются `index`, данные `data` могут быть любыми, name является именем `Series`.

In [4]:
data = ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog']
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
s = pd.Series(data, index=labels, name='animals')
s

a      cat
b      cat
c    snake
d      dog
e      dog
f      cat
g    snake
h      cat
i      dog
j      dog
Name: animals, dtype: object

### Создать `Dataframe` проще из словаря, а в качестве значений строк будем использовать список

In [3]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
         'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
         'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
         'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)
df

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

### А теперь сделаем открытие! Столбцы в `DataFrame` являются `Series`!

In [62]:
s = df.age
s

a    2.5
b    3.0
c    0.5
d    NaN
e    5.0
f    2.0
g    4.5
h    NaN
i    7.0
j    3.0
Name: age, dtype: float64

Обращение по индексу `s[0]`

In [63]:
s[0]

2.5

Элементы по определенным индексам `s[[4, 3, 1]]`

In [64]:
s[[4, 3, 1]]

e    5.0
d    NaN
b    3.0
Name: age, dtype: float64

Срезы `s[:3]`



In [65]:
s[:3]

a    2.5
b    3.0
c    0.5
Name: age, dtype: float64

Идеология словаря при обращении к элементам датафрейма

In [66]:
s['a']

2.5

In [67]:
'e' in s

True

In [68]:
't' in s

False

__Арифметические операции к Series__



In [69]:
s + s

a     5.0
b     6.0
c     1.0
d     NaN
e    10.0
f     4.0
g     9.0
h     NaN
i    14.0
j     6.0
Name: age, dtype: float64

__Логические операции к Series.__ Например, выбрать все значения, больше заданного



In [70]:
s[s > 3]

e    5.0
g    4.5
i    7.0
Name: age, dtype: float64

Функции из модуля numpy применяются поэлементно. Например, экспонента `np.exp(s)`

Количество строк можно узнать так `len(s)`

##   Редактирование таблиц. Создание и удаление столбцов. 

<div class="alert alert-info">
К столбцам `DataFrame` применимы любые математические операции. Надо толькo понимать, как работает векторизация.

Так, мы можем присвоить всему столбцу `'visits'` значение 3 и оно автоматические появится в каждой строке:

In [6]:
df['visits'] = 3
df

animal  age  visits priority
a    cat  2.5       3      yes
b    cat  3.0       3      yes
c  snake  0.5       3       no
d    dog  NaN       3      yes
e    dog  5.0       3       no
f    cat  2.0       3       no
g  snake  4.5       3       no
h    cat  NaN       3      yes
i    dog  7.0       3       no
j    dog  3.0       3       no

Надо увеличить возраст на год?  Просто увеличиваем столбец `'age'` на единицу. Даже не верится, что все так просто! 



In [7]:
df.age = df.age + 1
df

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
c  snake  1.5       3       no
d    dog  NaN       3      yes
e    dog  6.0       3       no
f    cat  3.0       3       no
g  snake  5.5       3       no
h    cat  NaN       3      yes
i    dog  8.0       3       no
j    dog  4.0       3       no

Можем даже создать новый столбец, используя другие столбцы `'age'` и `'visits'`. Казалось бы, всего одна строчка, но применяется для всего столбца! Векторизация это секретная магия pandas =)

Аналогично, можно  вычесть, умножить, разделить и сложить два численных столбца друг с другом.

In [8]:
df['age_visits_sum'] = df.age + df.visits
df

animal  age  visits priority  age_visits_sum
a    cat  3.5       3      yes             6.5
b    cat  4.0       3      yes             7.0
c  snake  1.5       3       no             4.5
d    dog  NaN       3      yes             NaN
e    dog  6.0       3       no             9.0
f    cat  3.0       3       no             6.0
g  snake  5.5       3       no             8.5
h    cat  NaN       3      yes             NaN
i    dog  8.0       3       no            11.0
j    dog  4.0       3       no             7.0

<div class="alert alert-info">
    
<b>Важное замечание:</b>
При создании новой колонки рекомендуется писать её название без пробелов (хотя с ними тоже можно) - как в примере выше: `age_visits_sum`.
Тогда можно будет обращаться к ней не только как ключу словаря, но и как к параметру датафрейма через точку.<br>
**`df.age_visits_sum`**

In [9]:
# Например
df.age_visits_sum.value_counts()

7.0     2
6.5     1
4.5     1
9.0     1
6.0     1
8.5     1
11.0    1
Name: age_visits_sum, dtype: int64

Также можем и удалить столбец, не забываем про `inplace=True`.

In [10]:
df.drop('age_visits_sum', axis=1, inplace=True)
df

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
c  snake  1.5       3       no
d    dog  NaN       3      yes
e    dog  6.0       3       no
f    cat  3.0       3       no
g  snake  5.5       3       no
h    cat  NaN       3      yes
i    dog  8.0       3       no
j    dog  4.0       3       no

<div class="alert alert-info">
<b>Полезная информация</b>


- __Удаление столбцов или строк__ `df.drop(name, axis=0, inplace=True)`  - Тут name это индекс строки или название колонки. Параметр `axis=0` означает, что удаляется строка, для удаления колонок (столбцов) нужно писать вместо индекса имена колонок списком и ставить `axis=1`.

    
- __Удаление дубликатов строк__ `df.drop_duplicates()` и не забываем про параметр `inplace=True`. Если надо удалить дубли по конкретному столбцу, что парадаем в метод название или список названий. 

    
    
- __Вставка в заданное место столбца__ `df.insert(loc, column, value)`.  Тут `loc` (куда), `column` (название), `value` (что вставлям). По умолчанию столбцы вставляются в конец.

    
    
- __Переименовать столбец__ можно так `df.rename(columns = {'old_name':'new_name'}, inplace=True)`</div>


# <center> 🔍 Разведочный анализ данных (EDA) </center>

<p id="2"></p>

В этой части непосредственно познакомимся с загрузкой данных, первичным осмотром и анализом.

 <img src="https://sun9-46.userapi.com/impg/4yA0_8N5ksJzftOvHsi_niXDlu4Ocy-HyjVj0w/YBBKjZhZsjc.jpg?size=1600x666&quality=96&sign=a6b10953218cb38c8e3042879b20935b&type=album"></img>

## Загрузка и первичное знакомство

<p id="3"></p>
<div class="alert alert-info">

**[pd.read_csv()](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)** - cчитывает данные в DataFrame из файла csv (файл с разделителями). У метода очень много параметров с которыми нужно ознакомится самостоятельно. 


In [ ]:
path = '../data/demo_rides_info.csv'
rides_info = pd.read_csv(path)
rides_info.head(10)

NameError: name 'pd' is not defined

<div class="alert alert-info"><b>Полезная информация</b>

- Если при чтении файла вы видите столцы с названием типа `Unnamed: 0` , то это сохраненный индекс при записи файла. Можно загрузить индексы из этого столбца не создавая их по умолчанию с нуля с помощью параметра  `index_col=0`, указав номер столбца который нужно использовать для индексов. 

    
- Полезные методы вывода строк: `df.head(n)` / `df.sample(n)` / `df.tail(n)` - возвращает первые, случайные или последние n строк датафрейма. Структура `DataFrame` А что это? `DataFrame` – это двумерная структура данных со столбцами потенциально разных типов.

    
- Подсчет числа каждого значения:  `df.value_counts()` - функция используется для получения уникальных значений и числа их встречаемости в виде `Series` . По умолчанию отсортирован по убыванию встречаемости `(ascending=False)`. Можно изменить сортировку через `ascending=True`. Также числовые столбцы можно разбивать на бины через параметр `bins`.  

</div>

## Предварительный осмотр данных

<p id="4"></p>

**[df.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html)** - возвращает случайную выборку строк датафрейма. Параметр n - кол-во выводимых строк.  

In [76]:
rides_info.sample(5)

user_id      car_id ride_id   ride_date  rating  ride_duration  \
12045  U17038085p  y31807857l     V1m  2020-03-16    2.64            106   
40839  k10615711l  x95423952G     M1Z  2020-01-30    5.75             43   
11020  m16402716a  d-1431691v     Z1d  2020-03-07    0.10             51   
15050  A38938958p  s-2436543G     t1s  2020-01-04    5.57            160   
48638  G14384502a  j21109696Z     A1m  2020-02-19    3.80            522   

       ride_cost  speed_avg   speed_max  stop_times     distance  refueling  \
12045       1690         47   75.000000           2   285.568460          0   
40839        425         70  111.188451           0  2996.384525          0   
11020        606         69  141.000000           1  3592.916070          0   
15050       1273         36   49.000000          17  1990.118048          0   
48638       7302         53   66.000000           1    97.144112          0   

       user_ride_quality  deviation_normal  
12045          -9.140320           -10.710  
40839          -4.831653           -38.001  
11020          -2.858706            -2.607  
15050           3.964930             7.409  
48638         -22.225319           -25.942

**[df.head()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html)** - возвращает первые строки датафрейма. По умолчаю выводит 5 строк, но можно изменить параметр n и вывести больше.  

Часто используют метод `head()`, но лучше первый раз смотреть на датафрейм через `sample()`.
Так вы имеете возможность смотреть на разные выборки и это может быть полезно, чтобы увидеть какие-нибудь баги.

In [77]:
rides_info.head()

user_id      car_id ride_id   ride_date  rating  ride_duration  \
0  q53527020Q  g55775125g     i1O  2020-01-16    3.42             33   
1  X72952203K  Y-6420899e     Z1i  2020-01-19    4.00             65   
2  g40994147T  z77468044J     x1M  2020-01-21    6.18             46   
3  E18650750E  W50413234I     W1Y  2020-03-17    1.43             80   
4  D18083254C  e-4559210H     U1f  2020-03-21    3.45         142292   

   ride_cost  speed_avg   speed_max  stop_times      distance  refueling  \
0        324         75  100.298135           0  2.412860e+03          0   
1        839         51   78.000000           0  2.553284e+03          0   
2        363         47   73.000000           2  2.079931e+02          0   
3        955         34  116.000000           2  1.758479e+03          0   
4    2276667         52   64.000000           1  7.541414e+06          0   

   user_ride_quality  deviation_normal  
0          -3.828166           -12.577  
1          -0.166199            -5.850  
2          -6.904072             0.000  
3           3.132158            18.966  
4         -29.248641            -8.126

<div class="alert alert-info">
    <b>Выводы</b>
    <ul>
    <li>в целом данные загрузились</li>
    <li>названия столбцов из первой строки загружены</li>
    <li>индексы первого столбца выглядят адекватно</li>
    <li>наличие индексов, которые нужно удалить не обнаружено</li>
    <li>по случайным пяти строкам наличие каких-либо ошибок не обнаружено</li></ul>
</div>

<div class="alert alert-info">

**[df.info()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html)*** - Выводит краткую сводку информацию по датафрейму.  

В нашем случае датафрейм небольшой, но он может быть большим. Мы разберем методы библиотеки Pandas. на этом небольшом примере, которые вы по аналогии самостоятельно сможете применить к другим датасетам

In [78]:
rides_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   user_id            50000 non-null  object 
 1   car_id             50000 non-null  object 
 2   ride_id            50000 non-null  object 
 3   ride_date          50000 non-null  object 
 4   rating             50000 non-null  float64
 5   ride_duration      50000 non-null  int64  
 6   ride_cost          50000 non-null  int64  
 7   speed_avg          50000 non-null  int64  
 8   speed_max          49747 non-null  float64
 9   stop_times         50000 non-null  int64  
 10  distance           50000 non-null  float64
 11  refueling          50000 non-null  int64  
 12  user_ride_quality  49841 non-null  float64
 13  deviation_normal   50000 non-null  float64
dtypes: float64(5), int64(5), object(4)
memory usage: 5.3+ MB


<div class="alert alert-info"><b>Выводы</b>

* в датафрейме 14 столбцов
* четыре из них имеют тип `object`
* пять из них имеют тип `float64`
* пять столбцов имеют тип `int64`
* всего в датафрейме 50000 строк
* в столбцах `speed_max` и `user_ride_quality` есть пропуски (nan)
* объем занимаемый датафреймом 5.3+ MB (эта информация может быть полезна, чтобы заранее представить какие операции с датафреймом вы сможете выполнять с учетом доступной вам оперативной памяти)


<div class="alert alert-info"><b>Выводы</b>

**[df.columns](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.columns.html)*** - Выводит метки столбцов датафрейма.  

<div class="alert alert-info"><b>Выводы</b>
Иногда кол-во столбцов очень большое, так что при выводе они все не помещаются в стандартный вывод через `df.info()`. Тогда можно воспользоваться атрибутом для просмотра индексов названий столбцов.

In [79]:
rides_info.columns

Index(['user_id', 'car_id', 'ride_id', 'ride_date', 'rating', 'ride_duration',
       'ride_cost', 'speed_avg', 'speed_max', 'stop_times', 'distance',
       'refueling', 'user_ride_quality', 'deviation_normal'],
      dtype='object')

Гораздо нагляднее и удобнее работать со списком названий столбцов

In [80]:
column_names = list(rides_info.columns)
column_names, len(column_names)

(['user_id',
  'car_id',
  'ride_id',
  'ride_date',
  'rating',
  'ride_duration',
  'ride_cost',
  'speed_avg',
  'speed_max',
  'stop_times',
  'distance',
  'refueling',
  'user_ride_quality',
  'deviation_normal'],
 14)

<div class="alert alert-info"><b>Выводы</b>
<ul><li>14 столбцов</li>
<li>все названия столбцов в нижнем регистре и не содержат спецсимволов (значит их не нужно переименовывать, чтобы обращаться к ним через точку)</li></ul></div>


## Работа с числовыми данными

<div class="alert alert-info">
<p id="5"></p>

**[df.describe()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html)*** - Выводит описательные статистики по числовым столбцам датафрейма. Описательные статистики включают в себя
- `count` - кол-во не пустых значений
- `mean` - среднее арифметическое
- `std` - дисперсия
- `min`, `max` - минимум и максимум
- 25%, 50%, 75% - соотвествующие квартили  


__!!!Bнимание!!!__ из статистик автоматически исключаюися `NaN` значения, если они есть.  

Выведем такое описание таблицы, округлив значения до второго знака после запятой:

In [81]:
rides_info.describe().round(2)

rating  ride_duration    ride_cost  speed_avg  speed_max  stop_times  \
count  50000.00       50000.00     50000.00   50000.00   49747.00    50000.00   
mean       4.49        9304.08    117999.44      47.10      77.48        1.34   
std        2.13       35947.59    504330.21      12.65      24.28        2.35   
min        0.00           2.00         9.00      25.00      30.00        0.00   
25%        3.14          27.00       297.00      38.00      62.00        0.00   
50%        4.49          44.00       506.00      46.00      72.00        1.00   
75%        5.86          69.00       895.00      52.00      86.00        2.00   
max       10.00      222356.00  10889545.00     100.00     202.46       21.00   

          distance  refueling  user_ride_quality  deviation_normal  
count     50000.00    50000.0           49841.00          50000.00  
mean     444872.13        0.0              -0.19             -1.20  
std     1782500.58        0.0              11.79             19.40  
min           3.04        0.0             -59.52           -110.76  
25%         784.16        0.0              -6.94             -9.01  
50%        1456.72        0.0               0.16              0.00  
75%        2265.27        0.0               6.69              6.98  
max    18333880.69        1.0              56.53             99.92

<div class="alert alert-info">
    <b>Выводы</b>
    
- описание столбцов датафрейма (`speed_avg` - средняя скорость по маршруту такси, `rating` - рейтинг, полученный водителем за поездку, `ride_duration` - продолжительность поездки, `ride_cost` - стоимость поездки, `stop_times` - кол-во остановок)
- средняя скорость движения по маршруту по всем поездкам около 40, а максимальная средняя скорость 103. Вероятно этот столбец имеет размерность км/час
- как минимум в половине случаев во время поездки была одна остановка
- максимальный рейтинг около 8, минимум чуть ниже единицы. Более половины всех рейтингов с оценками от 3 до 5. Вероятно размерность этого столбца баллы
- сделать однозначные выводы о размерностях продолжительности и стоимости поездок нельзя, вероятно они были модифицированы и для этого требуется дополнительный анализ и визуализация данных   

</div>

`describe()` так же можно использовать и для категориальных столбцов, добавив параметр `include='object'`

In [36]:
rides_info.describe(include='object')

user_id      car_id ride_id   ride_date
count        50000       50000   50000       50000
unique       13975        4250    2704          92
top     Z20970325i  a11867112k     H1v  2020-01-16
freq            18          25      34         604

***[df.value_counts()](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html)*** - функция используется для получения уникальных значений и числа их встречаемости в виде Series . По умолчанию отсортирован по убыванию встречаемости (`ascending=False`). Можно изменить сортировку через `ascending=True`. Также числовые столбцы можно разбивать на бины через параметр bins.  



In [82]:
rides_info.ride_id.value_counts()[:5]

H1v    34
E1y    34
K1F    33
H1z    33
P1Q    33
Name: ride_id, dtype: int64

Вывод: самый часто встречающийся код маршрута H1v и E1y = 34 поездки 

Посмотрим как распределяется кол-во поездок по рейтингу поездок разбивая его на бины:

In [83]:
rides_info.rating.value_counts(bins=4)

(2.5, 5.0]       21730
(5.0, 7.5]       15101
(-0.011, 2.5]     8761
(7.5, 10.0]       4408
Name: rating, dtype: int64

Вывод: 26% (8761+4408 делить на 18000) поездок с рейтингом ниже 2.5 и выше 7.5



**[.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html)*** - метод который позволит нам выявить дублирование в данных. Возвращает булевы значения, обозначающие повторяющиеся строки. При этом можно не учитывать некоторые определенные столбцы.

In [133]:
df = pd.DataFrame({
    'brand': ['Yum Yum', 'Yum Yum', 'Indomie', 'Indomie', 'Indomie'],
    'style': ['cup', 'cup', 'cup', 'pack', 'pack'],
    'rating': [4, 4, 3.5, 15, 5]
})
df.duplicated()

0    False
1     True
2    False
3    False
4    False
dtype: bool

Также можно использовать на определенном столбце/столбцах.

In [134]:
df.duplicated(subset=['brand'])

0    False
1     True
2    False
3     True
4     True
dtype: bool

# <center> 📤 Извлечение необходимой информации </center>

<p id="6">Перейдем к самому интересному - непосредственно к анализу полученных данных, познакомимся с методами сортировки, фильтрации и разными функциями.</p>

<img src="https://sun9-41.userapi.com/impg/io095r0KS1Fq4NTHYReNuRTf6DBM1pihDZMPqQ/Xki0IFcFAK8.jpg?size=2536x1150&quality=96&sign=db8f9c09f39dd1417032d06830270d44&type=album"></img>

## Сортировка DataFrame

<div class="alert alert-info">
<p id="7">
    
`DataFrame` можно сортировать. Посмотрим как это можно делать.</p>


__По значениям__

In [84]:
df.sort_values(by='age') 

animal  age  visits priority
c  snake  1.5       3       no
f    cat  3.0       3       no
a    cat  3.5       3      yes
b    cat  4.0       3      yes
j    dog  4.0       3       no
g  snake  5.5       3       no
e    dog  6.0       3       no
i    dog  8.0       3       no
d    dog  NaN       3      yes
h    cat  NaN       3      yes

__По индексу__

In [85]:
df.sort_index(axis=1, ascending=False)

visits priority animal  age
a       3      yes    cat  3.5
b       3      yes    cat  4.0
c       3       no  snake  1.5
d       3      yes    dog  NaN
e       3       no    dog  6.0
f       3       no    cat  3.0
g       3       no  snake  5.5
h       3      yes    cat  NaN
i       3       no    dog  8.0
j       3       no    dog  4.0

<div class="alert alert-info">
    
Параметр `ascending=True` по умолчанию, соответственно сортируется по возрастанию. Если нужно сортировать по нескольким столбцам, то передается список столбов и список булевых переменных в `ascending`. Параметр `inplace=False` по умолчанию.



## Фильтрация данных

<p id="8">Логические выражения используются для получения подтаблицы с заданными условиями по столбцам.</p>

Рассмотрим примеры.



In [86]:
df.loc[df.visits > 2]

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
c  snake  1.5       3       no
d    dog  NaN       3      yes
e    dog  6.0       3       no
f    cat  3.0       3       no
g  snake  5.5       3       no
h    cat  NaN       3      yes
i    dog  8.0       3       no
j    dog  4.0       3       no

Если условий больше одного, то условия фильтрации записываются в круглых скобках!

- __'И'__ обозначается как __&__
- __'ИЛИ'__  обозначается как __|__

In [87]:
df.loc[(df.visits > 2) & (df.priority == 'yes')]

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
d    dog  NaN       3      yes
h    cat  NaN       3      yes

In [88]:
df.loc[(df.visits > 2) | (df.age > 5)]

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
c  snake  1.5       3       no
d    dog  NaN       3      yes
e    dog  6.0       3       no
f    cat  3.0       3       no
g  snake  5.5       3       no
h    cat  NaN       3      yes
i    dog  8.0       3       no
j    dog  4.0       3       no

Чтобы нам вывести часть датафрейма с дешевыми ценами (цены поездки ниже 500), попробуем просто записать это условие и вывести часть датафрейма, которая соответствует этому условию 

In [89]:
rule_cheap = rides_info.ride_cost < 500
rides_info[rule_cheap].head().round(2)

user_id      car_id ride_id   ride_date  rating  ride_duration  \
0   q53527020Q  g55775125g     i1O  2020-01-16    3.42             33   
2   g40994147T  z77468044J     x1M  2020-01-21    6.18             46   
6   j17924049w  g10190493K     e1F  2020-03-15    3.80             33   
11  N20390973B  T12315854s     g1g  2020-03-18    8.59             27   
12  x37836441s  o-5998359x     W1M  2020-01-05    7.88             14   

    ride_cost  speed_avg  speed_max  stop_times  distance  refueling  \
0         324         75     100.30           0   2412.86          0   
2         363         47      73.00           2    207.99          0   
6         426         56     121.03           0   1854.43          0   
11        372         48      66.00           1   1061.04          0   
12        177         38      46.00           2    470.41          0   

    user_ride_quality  deviation_normal  
0               -3.83            -12.58  
2               -6.90              0.00  
6              -21.87            -20.00  
11              -9.69            -18.86  
12               7.31              2.39

Видно, что в выборку попали строки с индексом 0 и 2 (значения строк в столбце ride_cost 324 и 363 соответственно ).


### Булева маска

Также одним из самых простых способов отфильтровать нужные строки, это добавить булевую маску. Пусть маска строится по условию, что стоимость поездки больше некоторого значения.

- Создаем булеву маску  `[True, False, True, ...]`
- Выводим первые 5 значений строк, подходящих под условие

In [90]:
rule_cheap = rides_info['ride_cost'] < 426  
rides_info[rule_cheap].head()      

user_id      car_id ride_id   ride_date  rating  ride_duration  \
0   q53527020Q  g55775125g     i1O  2020-01-16    3.42             33   
2   g40994147T  z77468044J     x1M  2020-01-21    6.18             46   
11  N20390973B  T12315854s     g1g  2020-03-18    8.59             27   
12  x37836441s  o-5998359x     W1M  2020-01-05    7.88             14   
16  v23018580U  J-9977061N     e1v  2020-02-18    5.01             30   

    ride_cost  speed_avg   speed_max  stop_times     distance  refueling  \
0         324         75  100.298135           0  2412.859858          0   
2         363         47   73.000000           2   207.993094          0   
11        372         48   66.000000           1  1061.035898          0   
12        177         38   46.000000           2   470.414162          0   
16        415         35   80.000000           0  1045.715427          0   

    user_ride_quality  deviation_normal  
0           -3.828166           -12.577  
2           -6.904072             0.000  
11          -9.692375           -18.856  
12           7.314068             2.389  
16          -5.538496            18.364

Если мы хотим добавить новый столбец `cheap_price` - характеристика дешевой цены поездки: если цена в первом 25% квартиле (то есть меньше 426) то это будет дешевая цена `Yes`, иначе Nan.

In [91]:
rule_cheap = rides_info.ride_cost < 426
rides_info.loc[rule_cheap, 'cheap_price'] = 'Yes'
rides_info.head().round(2)

user_id      car_id ride_id   ride_date  rating  ride_duration  \
0  q53527020Q  g55775125g     i1O  2020-01-16    3.42             33   
1  X72952203K  Y-6420899e     Z1i  2020-01-19    4.00             65   
2  g40994147T  z77468044J     x1M  2020-01-21    6.18             46   
3  E18650750E  W50413234I     W1Y  2020-03-17    1.43             80   
4  D18083254C  e-4559210H     U1f  2020-03-21    3.45         142292   

   ride_cost  speed_avg  speed_max  stop_times    distance  refueling  \
0        324         75      100.3           0     2412.86          0   
1        839         51       78.0           0     2553.28          0   
2        363         47       73.0           2      207.99          0   
3        955         34      116.0           2     1758.48          0   
4    2276667         52       64.0           1  7541414.25          0   

   user_ride_quality  deviation_normal cheap_price  
0              -3.83            -12.58         Yes  
1              -0.17             -5.85         NaN  
2              -6.90              0.00         Yes  
3               3.13             18.97         NaN  
4             -29.25             -8.13         NaN

### Методы `.filter()` и `.query()`


Выбор строк или столбцов



In [92]:
df.filter(items=['a', 'j'], axis=0)

animal  age  visits priority
a    cat  3.5       3      yes
j    dog  4.0       3       no

<div class="alert alert-info">

__Пояснение__

Этот метод также имеет параметр `like`. Он отбирает строки / столбцы, в названии которых есть подстрока, присвоенная ему. Параметры `items` и `like` взаимоисключающие.



In [ ]:
df.filter(like='a', axis=1)

__Фильтрация__ альтернативным способом



In [94]:
df.query('age > visits')

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
e    dog  6.0       3       no
g  snake  5.5       3       no
i    dog  8.0       3       no
j    dog  4.0       3       no

<div class="alert alert-info">

__Пояснение__

Если в запросе необходимо использовать переменную, например, `mean_age` то перед ней пишем `@`

`df.query('age > @mean_age')`

Можно использовать сложные запросы с __'И'__ обозначается как `&`, __'ИЛИ'__  обозначается как `|`.

`df.query('age > @mean_age | animal == 'cat')`




### Метод [`concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) позволяет объединять объекты pandas вдоль определенной оси.

In [130]:
s1 = pd.Series(['a', 'b'])
s2 = pd.Series(['c', 'd'])
pd.concat([s1, s2])

0    a
1    b
0    c
1    d
dtype: object

Также есть возможность задавать индексы и лэйблы.

In [131]:
pd.concat([s1, s2], keys=['s1', 's2'],
          names=['Series name', 'Row ID'])

Series name  Row ID
s1           0         a
             1         b
s2           0         c
             1         d
dtype: object

Пример соединения `DataFrame`.

In [132]:
df1 = pd.DataFrame([['a', 1], ['b', 2]],
                   columns=['letter', 'number'])
df2 = pd.DataFrame([['c', 3], ['d', 4]],
                   columns=['letter', 'number'])

pd.concat([df1, df2])

letter  number
0      a       1
1      b       2
0      c       3
1      d       4

### __Метод `isin()`__ - это один из самых полезных способов создать булевую маску для фильтрации

In [13]:
car2stay = ['g55775125g', 'Y-6420899e', 'z77468044J']
mask = rides_info['car_id'].isin(car2stay)

In [14]:
rides_info[mask]

user_id      car_id ride_id   ride_date  rating  ride_duration  \
0      q53527020Q  g55775125g     i1O  2020-01-16    3.42             33   
1      X72952203K  Y-6420899e     Z1i  2020-01-19    4.00             65   
2      g40994147T  z77468044J     x1M  2020-01-21    6.18             46   
5564   Y10710002e  Y-6420899e     K1s  2020-02-28    1.70            135   
8545   K18258387y  g55775125g     Q1N  2020-02-08    4.49          60889   
10177  b23053441W  z77468044J     v1C  2020-03-18    5.04             38   
11301  l14008982F  g55775125g     Z1T  2020-01-14    6.47            102   
11691  s88098975e  Y-6420899e     b1r  2020-01-13    5.78             11   
13076  E12674591q  Y-6420899e     K1l  2020-02-01    8.63             37   
13746  l84330297P  z77468044J     S1C  2020-01-13    3.53             42   
14811  O28287672L  g55775125g     x1v  2020-03-29    1.96             58   
15401  Y17253493s  z77468044J     w1w  2020-02-22    4.36          36038   
18347  v20596053B  Y-6420899e     y1Z  2020-01-29    4.36             34   
18473  j22908494H  g55775125g     s1V  2020-03-16    4.22         183144   
20814  Y17571504l  Y-6420899e     V1i  2020-03-14    5.98            230   
21214  Z50320089I  Y-6420899e     Y1F  2020-01-14    5.78             28   
21712  e33207700p  Y-6420899e     S1t  2020-01-10   10.00             14   
22105  k70641968u  Y-6420899e     t1C  2020-03-02    1.58            224   
22243  l53852012F  z77468044J     g1f  2020-01-16    3.85             36   
28272  A60966875K  Y-6420899e     L1K  2020-01-27    3.05             48   
29026  t21259369l  g55775125g     b1K  2020-03-28    2.29             18   
32584  J51636842i  g55775125g     y1Q  2020-03-15    0.10             46   
33276  t18687398C  Y-6420899e     w1w  2020-03-01    3.50             43   
33829  N84968544j  z77468044J     p1H  2020-03-25    2.58              6   
34866  O18712245E  g55775125g     w1u  2020-01-27    0.69             19   
35510  P71461395r  g55775125g     w1z  2020-02-07    7.11             47   
36270  Q21750842l  z77468044J     Y1h  2020-01-28    4.49             23   
36539  h18623490m  g55775125g     s1q  2020-02-22    2.47             19   
38120  U11342208s  z77468044J     r1p  2020-01-16    3.55             24   
40339  X19477026l  z77468044J     g1W  2020-02-29    5.56             31   
40713  o44394795h  Y-6420899e     m1l  2020-03-03    6.30             22   
42233  F62694314E  g55775125g     U1J  2020-03-14    4.92             66   
43119  o41298704Y  z77468044J     P1p  2020-03-19    5.54             70   
43535  S14629356D  z77468044J     n1z  2020-03-06    6.85         207426   
43687  k15218712c  Y-6420899e     C1S  2020-01-28    3.41             46   
44212  c22382731Y  g55775125g     c1k  2020-02-28    2.52             23   
44644  Z16377989S  Y-6420899e     v1o  2020-01-29    4.36             40   
44748  Q22804481u  Y-6420899e     I1N  2020-01-06    6.29             12   
44855  h92996794Y  z77468044J     O1I  2020-01-21    0.48             22   
47486  L16642682D  Y-6420899e     Y1m  2020-01-09    4.64             32   
48162  i44601400V  z77468044J     Z1C  2020-01-15    2.87             40   

       ride_cost  speed_avg   speed_max  stop_times      distance  refueling  \
0            324         75  100.298135           0  2.412860e+03          0   
1            839         51   78.000000           0  2.553284e+03          0   
2            363         47   73.000000           2  2.079931e+02          0   
5564        1884         64   66.336795           0  3.233434e+03          0   
8545      852440         57   66.920236           0  3.587582e+06          0   
10177        527         40   66.000000           1  1.061336e+03          0   
11301       1319         65   96.413232           0  1.833097e+03          0   
11691        148         43   48.000000           1  2.805963e+02          0   
13076        512         50   72.387594           0  1.899176e+03          0   
13746        415 

In [15]:
rides_info[~mask]

user_id      car_id ride_id   ride_date  rating  ride_duration  \
3      E18650750E  W50413234I     W1Y  2020-03-17    1.43             80   
4      D18083254C  e-4559210H     U1f  2020-03-21    3.45         142292   
5      G12536963g  B61056692H     f1p  2020-02-26    5.73             49   
6      j17924049w  g10190493K     e1F  2020-03-15    3.80             33   
7      Z21861082q  U-8316679F     s1l  2020-02-22    2.88             23   
...           ...         ...     ...         ...     ...            ...   
49995  g11939886D  I-9069619f     X1x  2020-03-28    3.90             30   
49996  P13881472G  T-4233717D     v1s  2020-02-10    2.83         202887   
49997  i60157442b  U11495954u     l1U  2020-02-09    7.60             47   
49998  a57837499I  G-5803601t     N1r  2020-01-25    7.16              7   
49999  w14744581R  O-1023170l     r1v  2020-01-24    6.03             56   

       ride_cost  speed_avg   speed_max  stop_times      distance  refueling  \
3            955         34  116.000000           2  1.758479e+03          0   
4        2276667         52   64.000000           1  7.541414e+06          0   
5            681         36   65.000000           1  1.633555e+03          0   
6            426         56  121.029522           0  1.854434e+03          0   
7            685         53   58.000000           0  1.290301e+03          0   
...          ...        ...         ...         ...           ...        ...   
49995        384         31   61.000000           0  9.726349e+02          0   
49996    1623091         43  117.316900           1  8.927038e+06          0   
49997        369         66   91.763722           0  3.150356e+03          0   
49998         78         34   67.000000           1  1.197659e+02          0   
49999        612         50   88.000000           1  2.302677e+03          0   

       user_ride_quality  deviation_normal  
3               3.132158            18.966  
4             -29.248641            -8.126  
5               5.566720             0.000  
6             -21.874094           -19.996  
7               1.181858            -0.799  
...                  ...               ...  
49995         -24.883368           -20.119  
49996           6.311743            -7.842  
49997          -9.265198             9.674  
49998         -11.737839             4.371  
49999           9.834685             4.280  

[49959 rows x 14 columns]

## Работа с пропущенными значениями

<p id="9"></p>

Pandas для представления отсутствующих данных использует значение из numpy `np.nan`. По умолчанию оно не включается в вычисления. Где хранятся `np.nan`  показывает `df.isna()` оно же `df.isnull()`



In [95]:
df.isnull()

animal    age  visits  priority
a   False  False   False     False
b   False  False   False     False
c   False  False   False     False
d   False   True   False     False
e   False  False   False     False
f   False  False   False     False
g   False  False   False     False
h   False   True   False     False
i   False  False   False     False
j   False  False   False     False

Метод `dropna()` удаляет NaN's из датасета

In [96]:
df.dropna()

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
c  snake  1.5       3       no
e    dog  6.0       3       no
f    cat  3.0       3       no
g  snake  5.5       3       no
i    dog  8.0       3       no
j    dog  4.0       3       no

<div class="alert alert-info">

__Замечание__

По умолчанию некоторые параметры - этого метода следующие:

- `axis=0`  0 удаляет строки, содержащие пропущенные значения, 1 удаляет столбцы.
- `how='any'` `'any'` если хотя бы одно `NA`, удаляет эту строку или столбец, `'all'` только если все значения `NA`
- `inplace=False` False выполняет операцию и возвращает копию объекта, True выполняет операцию на месте и возвращает `None`


Вжух, и все пропущенные значения исчезли! Однако, не забывайте про параметр `inplace=True` иначе ваши изменения не сохранятся. 



Заполняет пропущенные значения `df.fillna(n)`, при n = 5 имеем:

In [97]:
df.fillna(5)

animal  age  visits priority
a    cat  3.5       3      yes
b    cat  4.0       3      yes
c  snake  1.5       3       no
d    dog  5.0       3      yes
e    dog  6.0       3       no
f    cat  3.0       3       no
g  snake  5.5       3       no
h    cat  5.0       3      yes
i    dog  8.0       3       no
j    dog  4.0       3       no

<div class="alert alert-info">
    
По умолчанию некоторые параметры этого метода следующие

- `axis=None` 0 заполняет по строкам, 1 заполняет по столбцам

- `inplace=False` False выполняет операцию и возвращает копию объекта, True выполняет операцию на месте и возвращает None





Количество пропусков в каждом столбце

In [98]:
df.isna().sum()

animal      0
age         2
visits      0
priority    0
dtype: int64

`df.fillna()` - Заполняет значения NA/NaN. Параметр `inplace=True` заполняет в датафрейме без необходимости копирования. По умолчанию `inplace=False`.


## Основы индексации данных


<p id="10"></p>

**[df.loc[]](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html)** - Обеспечивает доступ к группе строк и столбцов по меткам (индексам) или логическому массиву. Обратите внимание, что, в отличие от обычных срезов Python, включены как начало, так и конец индексов. Через этот метод можно менять значения датафрейма.

<img src="https://sun2.beeline-kz.userapi.com/impg/I6sVsItmqvdhZODJ94ZFdwfvlYN3MMl7f5mDwQ/XmpQ6c34FX0.jpg?size=1646x732&quality=96&sign=1fe5b434040e4ad6cd7265352d10a4a3&type=album"></img>

Также, выбрать столбец можно как `df.col`, где col это имя столбца, если имя столбца написано латиницей и не содержит пробелов. На примерах это будет выглядеть так:  

In [99]:
df.animal

a      cat
b      cat
c    snake
d      dog
e      dog
f      cat
g    snake
h      cat
i      dog
j      dog
Name: animal, dtype: object

In [100]:
df.loc['a', ['animal', 'visits']]

animal    cat
visits      3
Name: a, dtype: object

In [101]:
df.iloc[:,0]

a      cat
b      cat
c    snake
d      dog
e      dog
f      cat
g    snake
h      cat
i      dog
j      dog
Name: animal, dtype: object

In [21]:
df.iloc[[1, 3], [0]]

animal
b    cat
d    dog

<div class="alert alert-info">

Наглядно это выглядит так

<img src="https://sun9-79.userapi.com/impg/p0NrZ2Owypd4NbKQV4--0kZYi4PfNqk3E31URQ/7NTTharc5XA.jpg?size=301x129&quality=96&sign=36a4edc1d86f1b6618a80545f36cf4ef&type=album"></img>

__Разница между `iloc` и `loc` простая!__   `.loc`- возвращает строку по номеру, `.iloc` по индексу. <br>
***Важно:*** Ещё одно различие состоит в том, что при наличии числового индекса `.iloc()` не будет включать правую границу в срез, а `.loc()` будет.

Посмотрим на примере:

In [31]:
# Строка с индексом 3 и колонка 2 не включены в срез
df2.iloc[1:3, 0:2]

animal  age
1    cat  4.0
2  snake  1.5

In [34]:
# А здесь правые границы включены в срез
df2.loc[1:3, 'animal':'visits']

animal  age  visits
1    cat  4.0       3
2  snake  1.5       3
3    dog  NaN       3

### Фокусы с индексами


Сбрасывает индекс `df.reset_index()`, причем предыдущий индекс не удаляется, а сохраняется в новой колонке. <br>

In [103]:
df.reset_index()

index animal  age  visits priority
0     a    cat  3.5       3      yes
1     b    cat  4.0       3      yes
2     c  snake  1.5       3       no
3     d    dog  NaN       3      yes
4     e    dog  6.0       3       no
5     f    cat  3.0       3       no
6     g  snake  5.5       3       no
7     h    cat  NaN       3      yes
8     i    dog  8.0       3       no
9     j    dog  4.0       3       no

Чтобы удалить предыдущий индекс и сохранить изменения, нужно указать параметры `drop=True, inplace=True`. 

In [28]:
df2 = df.reset_index(drop=True)
df2.head()

animal  age  visits priority
0    cat  3.5       3      yes
1    cat  4.0       3      yes
2  snake  1.5       3       no
3    dog  NaN       3      yes
4    dog  6.0       3       no

Устанавливает индекс `df.set_index()`. Соответственно, берет существующий столбец и переводит его в индекс.



Заменяет индекс `df.reindex()`



In [104]:
df.reindex(['j', 'i', 'h', 0, 1, 2], axis=0, fill_value=5)

animal  age  visits priority
j    dog  4.0       3       no
i    dog  8.0       3       no
h    cat  NaN       3      yes
0      5  5.0       5        5
1      5  5.0       5        5
2      5  5.0       5        5

<p id="11"></p>

## Функции
Применение функции к данным по строкам / столбцам реализуется `df.apply(func, axis=0)`

<img src="https://sun9-70.userapi.com/impg/SfdR2m0D2E02P1xeDa8XJTr2DopfykEufD7Kcw/rQkknfzl800.jpg?size=216x398&quality=96&sign=b395742fb9db51dbeb0e124d17f82901&type=album"></img>

В примере методу `apply()` передается функция, которая будет применяться к данным. Стоит заметить, что значения типа 'str' тоже суммируются.

In [106]:
df.apply(np.sum, axis=0)

animal      catcatsnakedogdogcatsnakecatdogdog
age                                       35.5
visits                                      30
priority              yesyesnoyesnononoyesnono
dtype: object

Применение функции к данным поэлементно реализуется `df.applymap(lambda x: func)`

В примере методу `applymap()`  передается `lambda`-функция, которая будет применяться к элементу.

In [107]:
df.applymap(lambda x: len(str(x)))

animal  age  visits  priority
a       3    3       1         3
b       3    3       1         3
c       5    3       1         2
d       3    3       1         3
e       3    3       1         2
f       3    3       1         2
g       5    3       1         2
h       3    3       1         3
i       3    3       1         2
j       3    3       1         2

<div class="alert alert-info">

Так же функцию `apply()` можно применять для создания новых колонок на основе уже существующей или на основе всего датафрейма. <br>
При этом есть существенная разница при обращении к одной колонке или к всему датасету.<br>
Разберём на примере:

In [4]:
df.head()

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no

In [6]:
# Создадим бинарную колонку 'more_2', если было больше 2-х визитов на основе колонки 'visits'
df['more_2'] = df['visits'].apply(lambda x: 1 if x > 2 else 0)
df.head()

animal  age  visits priority  more_2
a    cat  2.5       1      yes       0
b    cat  3.0       3      yes       1
c  snake  0.5       2       no       0
d    dog  NaN       3      yes       1
e    dog  5.0       2       no       0

In [11]:
# теперь создадим такую же колонку с названием 'more_2v2' на основе всего df
df['more_2v2'] = df.apply(lambda x: 1 if x.visits > 2 else 0, axis=1)
df.head()

animal  age  visits priority  more_2  more_2v2
a    cat  2.5       1      yes       0         0
b    cat  3.0       3      yes       1         1
c  snake  0.5       2       no       0         0
d    dog  NaN       3      yes       1         1
e    dog  5.0       2       no       0         0

Отметим, что при использовании `apply()` к датасету - надо указывать название колонки через точку и параметр `axis`.

### Встроенные функции

Здесь перечислены некоторые полезные методы, применимые к Series

<img src="https://sun9-17.userapi.com/impg/QIvfQYVUFnOii25XfA6yL9HSsnVaMSI71Ccqow/BROCketQrDw.jpg?size=1302x1068&quality=96&sign=ffe48fa08279d37d7e023cce818f888c&type=album" width=700></img>

Например, посчитаем количество уникальных животных:

In [12]:
df.animal.nunique()

3

<div class="alert alert-info">
    
<b> Заключение: </b>
В этом уроке мы познакомились с основными возможностями библиотеки `Pandas`, для закрепления материала рекомендуем прорешать задачи в степах.<br>
С более продвинутыми функциями библиотеки и их применением для решения практических задач будем разбираться в следующих уроках.